### DISCUSSION TITLE
How to specify precise level?

### DISCUSSION THREAD

**Discussion by franz1981** [async-profiler/async-profiler#602] [Q&A]
https://github.com/torvalds/linux/blob/master/tools/perf/Documentation/perf-list.txt#L60 reports (if available) 3 precise level for PMU events: I got a use case with cpu events getting a huge skidding - is it possible to sample cpu events specifying the precise level?

I see in the source code that perf_event_attr.precise_ip is specified as 2 for SW counters (lowering the level if compared to the maximum ie 3) - but I cannot find where the settings for the others is happening

**Comment by apangin** [comment]
> I got a use case with cpu events getting a huge skidding

I'd be interested to see a case where setting `perf_event_attr.precise_ip` results in a better quality profile. So far I haven't seen a real example where `precise_ip` makes any difference, except that on some systems `perf_event_open` instantly fails when `precise_ip` is other than 0.

**Reply by apangin** [reply]
To expand the above thought:
 1. On some architectures/kernels, setting `precise_ip` to anything but zero does not work at all, hence the default value of 0.
 2. I could probably add an option to control `precise_ip` manually, but I need a test case that reliably proves the option works as expected.

**Comment by franz1981** [comment]
Thanks @apangin, TBH the use case is a server with load approaching 100% (based on number of samples, hence an approximation) - where the same method call appear both with and without any ancestor or children. And when children are present they report a large number of samples on methods that shouldn't cost that much (I am investigating).
I wanted to exclude the skidding issue by raising the precision and "see" what I would get.
If you suggest other approaches, I am happy to try it.
I can post the flamegraph if you are curious

**Reply by apangin** [reply]
Understood.
I assume you can just compile your own version of async-profiler with `attr.precise_ip = 3;` and see what happens, right? I did that experiment before but didn't see much difference. It could be a wrong test though, so please try on your example and let me know.
At the same time, I guess those inaccuracies could be rather caused by imprecise pc->bytecode mapping produced by JIT.

**Reply by apangin** [reply]
BTW, `precise_ip = 3` is applicable to `-e cycles`, not `-e cpu`.
